File: server/src/main/java/se/kth/id2203/HostComponent.java
/*
 * The MIT License
 *
 * Copyright 2017 Lars Kroll <lkroll@kth.se>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package se.kth.id2203;

import se.kth.id2203.networking.NetAddress;
import se.sics.kompics.Channel;
import se.sics.kompics.Component;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Init;
import se.sics.kompics.network.Network;
import se.sics.kompics.network.netty.NettyInit;
import se.sics.kompics.network.netty.NettyNetwork;
import se.sics.kompics.timer.Timer;
import se.sics.kompics.timer.java.JavaTimer;

/**
 *
 * @author Lars Kroll <lkroll@kth.se>
 */
public class HostComponent extends ComponentDefinition {

    final NetAddress self = config().getValue("id2203.project.address", NetAddress.class);
    protected final Component timer = create(JavaTimer.class, Init.NONE);
    protected final Component net = create(NettyNetwork.class, new NettyInit(self));
    protected final Component parent = create(ParentComponent.class, Init.NONE);

    {
        connect(timer.getPositive(Timer.class), parent.getNegative(Timer.class), Channel.TWO_WAY);
        connect(net.getPositive(Network.class), parent.getNegative(Network.class), Channel.TWO_WAY);
    }
}

File: src/main/scala/example/HostComponent.scala
package example

import java.net.InetAddress

import example.bootstrap.networking.NetAddress
import se.sics.kompics.Init
import se.sics.kompics.network.Network
import se.sics.kompics.network.netty.{NettyInit, NettyNetwork}
import se.sics.kompics.sl.ComponentDefinition
import se.sics.kompics.timer.Timer
import se.sics.kompics.timer.java.JavaTimer

class HostComponent extends ComponentDefinition {
//    val self = cfg.getValue[example.TAddress]("host.address") // TODO Define
    val self = new NetAddress(InetAddress.getByName("http://localhost"), 8082)

    val timer = create(classOf[JavaTimer], se.sics.kompics.Init.NONE)
    val network = create(classOf[NettyNetwork], new NettyInit(self))
    val parent = create(classOf[ParentComponent], Init.NONE)

    connect[Timer](timer -> parent)
    connect[Network](network -> parent)

}
